### PR TITLE
[Feature] long sequence prefill on ds-rc1

### PR DESCRIPTION
**Pull request by maksim-tsishkouski-epam**
### PR Category
Feature

### Summary
Issue https://github.com/tenstorrent/tt-metal/issues/35141
- Added model and mla chunking (not support for 8upr)
- Added chunked_flash_mla_prefill (first chunk will be processed with flash_mla_prefill as it's faster in current setup and needs less inter-device communication)
- Some pre-allocations and loop fuse in mla
- lm head by default now runs only on a single token, others are skipped
- Added C++ fixes needed for chunked_flash_mla_prefill from main

#### Logs from local tests

See [the comment](https://github.com/tenstorrent/tt-metal/pull/43812#issuecomment-4460795330)

#### chunked_flash_mla needed workaround
related to kv cache sharding - each device stores each user kv cache in case of 8upr (and 4 users in case of 32upr)
1. replicates query on each device (not column sharded heads)
2. calculates attn_out which is correct only on 1 device
3. then broadcasts in 2 stages
3.1. broadcast across column to have one correct row
3.2. calculate output (which is correct only for one row) and broadcast across row to have correct tensor on each device
image

#### QUAD metrics ([more info here](https://github.com/tenstorrent/tt-metal/issues/35141#issuecomment-4400276207))
| prefill length | prefill_t/s | prefill_time_to_token | Notes |
|---|---|---|---|
| 32k | 350.83 | 1.5 min | no model, no mla chunking |
| 48k | 308.12 | 2.6 min | no model, no mla chunking |
| 64k | 63.62 | 17 min | only mla chunking |
| 128k | 34.57 |  63 min | model and mla chunking |

### CODE REVIEW

**Comment by yieldthought** [comment]
Codex review:

I found two issues that should be fixed before merging:

1. [models/demos/deepseek_v3/tt/model/row_batched_model.py:427](https://github.com/tenstorrent/tt-metal/blob/2abb995442ac8d197ffa85b7f501816e14fc65b5/models/demos/deepseek_v3/tt/model/row_batched_model.py#L427) appends `hidden_for_mtp_chunk`, but that value is the list produced by starred unpacking (`[hidden_tensor]`), not the tensor itself. When MTP is enabled, generator.py calls this path with `return_hidden=True` at [models/demos/deepseek_v3/tt/generator.py:2478](https://github.com/tenstorrent/tt-metal/blob/2abb995442ac8d197ffa85b7f501816e14fc65b5/models/demos/deepseek_v3/tt/generator.py#L2478), so the first prefill returns `hidden_for_mtp = [[tensor], ...]`; [models/demos/deepseek_v3/tt/model/row_batched_model.py:435](https://github.com/tenstorrent/tt-metal/blob/2abb995442ac8d197ffa85b7f501816e14fc65b5/models/demos/deepseek_v3/tt/model/row_batched_model.py#L435) then passes a list of lists into `ttnn.concat`. That breaks MTP prefill before the MTP cache can be primed or `return_last_hidden` can be extracted. This should append the tensor (`hidden_for_mtp_chunk[0]`) or otherwise keep `hidden_for_mtp` as `list[ttnn.Tensor]`.

2. [models/demos/deepseek_v3/tt/model/row_batched_model.py:429](https://github.com/tenstorrent/tt-metal/blob/2abb995442ac8d197ffa85b7f501816e14fc65b5/models/demos/deepseek_v3/tt/model/row_batched_model.py#L429) concatenates chunk outputs on each mesh row before any row-wise composition. Each `_forward_prefill` chunk has already gone through `Embedding2D` reduce-scatter, so row N holds that row's slice for the current chunk. Concatenating locally produces `[chunk0 rowN, chunk1 rowN, ...]`, but the downstream full-output consumers still index as if each row owns one contiguous global sequence shard: for example [models/demos/deepseek_v3/tt/generator.py:2030](https://github.com/tenstorrent/tt-metal/blob/2abb995442ac8d197ffa85b7f501816e14fc65b5/models/demos/deepseek_v3/tt/generator.py#L2030) indexes host logits with `prompt_len - 1`, and [models/demos/deepseek_v3/tt/generator.py:2584](https://github.com/tenstorrent/tt-metal/blob/2abb995442ac8d197ffa85b7f501816e14fc65b5/models/demos/deepseek_v3/tt/generator.py#L2584) maps `prompt_len - 1` to `hidden_row_idx/full_seq_len`. For any multi-chunk prefill this selects logits/hidden for the wrong token (and MTP token/hidden alignment is wrong even after fixing the list issue). The full logits/hidden path needs to preserve global token order, or the MTP/host indexing needs to be made chunk-aware.

**Review by yieldthought — APPROVED** [review]

**Comment by maksim-tsishkouski-epam** [comment]
all_needed_local_tests logs

✅ [dual_demo_output.log](https://github.com/user-attachments/files/27803814/dual_demo_output.log)
✅ [dual_demo_stress_output.log](https://github.com/user-attachments/files/27803812/dual_demo_stress_output.log)
✅ [quad_test_model_long_prefill_single_galaxy_output.log](https://github.com/user-attachments/files/27803811/quad_test_model_long_prefill_single_galaxy_output.log)
✅ [quad_demo_output.log](https://github.com/user-attachments/files/27803809/quad_demo_output.log)
✅ [quad_demo_stress_output.log](https://github.com/user-attachments/files/27804659/quad_demo_stress_output.log)
✅  [dual_teacher_forced_output.log](https://github.com/user-attachments/files/27807442/dual_teacher_forced_output.log)
```
[1,0]<stdout>: XFAIL models/demos/deepseek_v3/demo/test_demo_teacher_forced.py::test_demo_teacher_forcing_accuracy[8-128-reference_file0] - User outputs diverged across batch (issue #35509). first mismatch at token 2: user0=671, user1=43
[1,0]<stdout>: XFAIL models/demos/deepseek_v3/demo/test_demo_teacher_forced.py::test_demo_teacher_forcing_accuracy[32-128-reference_file0] - User outputs diverged across batch (issue #35509). first mismatch at token 3: user0=3295, user1=3417
```
✅  [quad_teacher_forced_output.log](https://github.com/user-attachments/files/27808999/quad_teacher_forced_output.log)
```
[1,0]<stdout>: XFAIL models/demos/deepseek_v3/demo/test_demo_teacher_forced.py::test_demo_teacher_forcing_accuracy[8-128-reference_file0] - User outputs diverged across batch (issue #35509). first mismatch at token 3: user0=3295, user1=3417
[1,0]<stdout>: XFAIL models/demos/deepseek_v3/demo/test_demo_teacher_forced.py::test_demo_teacher_forcing_accuracy[32-128-reference_file0] - User outputs diverged across batch (issue #35509). first mismatch at token 39: user0=223, user1=343
```

**Review by pprajapatiTT — APPROVED** [review]
